From 93e7b2268fabc35d754ffe6389b5172b5917eb8c Mon Sep 17 00:00:00 2001
From: Guillem Jover <guillem@debian.org>
Date: Thu, 25 Apr 2024 22:44:19 +0200
Subject: [PATCH] src: Check whether SOURCE_DATE_EPOCH is set before parsing it

The dpkg-deb and dpkg-split program try to parse this environment
variable to use it for their timestamps inside files to generate
reproducible artifacts. But when the environment variable is set
but empty then the parsing function fails with a confusing error
message.

This is an issue when building a package directly via debian/rules
that uses the pkg-info.mk fragment file, because that one tries to
set the SOURCE_DATE_EPOCH and can end up setting it to an empty value
if the changelog contains an unfinished trailer. This is not an issue
when using dpkg-buildpackage, though because the code there will
fallback to use the current time if it there is no value from the
changelog.

Closes: #1069846
Based-on-patch-by: Rainer Weikusat <rweikusat@cyberadapt.com>
---
 src/deb/build.c   | 2 +-
 src/split/split.c | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/deb/build.c b/src/deb/build.c
index 1f0c050ee..16f2bafdf 100644
--- a/src/deb/build.c
+++ b/src/deb/build.c
@@ -597,7 +597,7 @@ do_build(const char *const *argv)
   m_output(stdout, _("<standard output>"));
 
   timestamp_str = getenv("SOURCE_DATE_EPOCH");
-  if (timestamp_str)
+  if (str_is_set(timestamp_str))
     timestamp = parse_timestamp(timestamp_str);
   else
     timestamp = time(NULL);
diff --git a/src/split/split.c b/src/split/split.c
index 771de626c..04d41b281 100644
--- a/src/split/split.c
+++ b/src/split/split.c
@@ -162,7 +162,7 @@ mksplit(const char *file_src, const char *prefix, off_t maxpartsize,
 	version = versiondescribe(&pkg->available.version, vdew_nonambig);
 
 	timestamp_str = getenv("SOURCE_DATE_EPOCH");
-	if (timestamp_str)
+	if (str_is_set(timestamp_str))
 		timestamp = parse_timestamp(timestamp_str);
 	else
 		timestamp = time(NULL);
-- 
2.43.0

